### PR TITLE
erts: Marginally shrink process structure

### PR DESCRIPTION
**Pull request by jhogberg**
The process structure is now 32 bytes smaller on 64-bit platforms. Most of it was accomplished by reordering but some changes in data sizes were made, hence the sweeping changes in the JIT.

### CODE REVIEW

File: erts/emulator/beam/code_ix.h [hunk 1]
```diff
@@ -87,7 +87,14 @@ typedef unsigned ErtsCodeIndex;
...
-    Uint  arity;
+
+#ifndef BEAMASM
+    BeamInstr arity;
+#else
+    /* The arity does not need to be instruction-aligned in the JIT and its
+     * maximum possible value is 255, so we can shave off a few bytes here. */
+    Uint32 arity;
+#endif
 } ErtsCodeMFA;
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply erts/emulator/beam/code_ix.h:98]: Good point :)

File: erts/emulator/beam/jit/beam_jit_common.cpp [hunk 2]
```diff
@@ -1106,11 +1106,11 @@ ErtsMessage *beam_jit_decode_dist(Process *c_p, ErtsMessage *msgp) {
 }
 
 /* Remove a (matched) message from the message queue. */
-Sint beam_jit_remove_message(Process *c_p,
-                             Sint FCALLS,
-                             Eterm *HTOP,
-                             Eterm *E,
-                             Uint32 active_code_ix) {
+int beam_jit_remove_message(Process *c_p,
+                            int FCALLS,
```
Comment (bjorng) [inline erts/emulator/beam/jit/beam_jit_common.cpp:1110]:
```suggestion
                            Sint32 FCALLS,
```

File: erts/emulator/beam/jit/beam_jit_common.hpp [hunk 3]
```diff
@@ -616,11 +616,11 @@ Eterm beam_jit_bs_get_integer(Process *c_p,
                               Uint Live);
 
 ErtsMessage *beam_jit_decode_dist(Process *c_p, ErtsMessage *msgp);
-Sint beam_jit_remove_message(Process *c_p,
-                             Sint FCALLS,
-                             Eterm *HTOP,
-                             Eterm *E,
-                             Uint32 active_code_ix);
+int beam_jit_remove_message(Process *c_p,
+                            int FCALLS,
```
Comment (bjorng) [inline erts/emulator/beam/jit/beam_jit_common.hpp:620]:
```suggestion
                            Sint32 FCALLS,
```

File: erts/emulator/beam/jit/x86/instr_bif.cpp [hunk 4]
```diff
@@ -572,7 +572,8 @@ void BeamGlobalAssembler::emit_call_light_bif_shared() {
         {
             a.mov(ARG2, mbuf_mem);
             a.mov(ARG5, export_mem);
-            a.mov(ARG5, x86::qword_ptr(ARG5, offsetof(Export, info.mfa.arity)));
+            a.mov(ARG5d,
+                  x86::dword_ptr(ARG5, offsetof(Export, info.mfa.arity)));
```
Comment (bjorng) [inline erts/emulator/beam/jit/x86/instr_bif.cpp:576]:
```suggestion
            a.movzx(ARG5d,
                  x86::byte_ptr(ARG5, offsetof(Export, info.mfa.arity)));
```

File: erts/emulator/beam/jit/x86/instr_bif.cpp [hunk 5]
```diff
@@ -609,9 +610,9 @@ void BeamGlobalAssembler::emit_call_light_bif_shared() {
 
     a.bind(yield);
     {
-        a.mov(ARG2, x86::qword_ptr(ARG4, offsetof(Export, info.mfa.arity)));
+        a.mov(ARG2d, x86::dword_ptr(ARG4, offsetof(Export, info.mfa.arity)));
```
Comment (bjorng) [inline erts/emulator/beam/jit/x86/instr_bif.cpp:613]:
```suggestion
        a.movzx(ARG2d, x86::byte_ptr(ARG4, offsetof(Export, info.mfa.arity)));
```

File: erts/emulator/beam/erl_bif_trace.c [hunk 6]
```diff
@@ -955,7 +955,7 @@ static Eterm
 trace_info_pid(Process* p, Eterm pid_spec, Eterm key)
 {
     Eterm tracer;
-    Uint trace_flags = am_false;
+    Uint32 trace_flags = 0;
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply erts/emulator/beam/erl_bif_trace.c:958]: Yes, it threw me off too but it seems like it's only ever used when it's explicitly updated to some flags and not an atom.

File: erts/emulator/beam/jit/arm/instr_bif.cpp [hunk 7]
```diff
@@ -539,7 +536,7 @@ void BeamGlobalAssembler::emit_call_light_bif_shared() {
             a.ldr(ARG2, mbuf_mem);
             load_x_reg_array(ARG4);
             a.ldr(ARG5, export_mem);
-            a.ldr(ARG5, arm::Mem(ARG5, offsetof(Export, info.mfa.arity)));
+            a.ldr(ARG5.w(), arm::Mem(ARG5, offsetof(Export, info.mfa.arity)));
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply erts/emulator/beam/jit/arm/instr_bif.cpp:539]: Oops, I forgot to change this to bytes after going from `Uint32`-sized arity to byte-sized.

File: erts/emulator/beam/code_ix.h [hunk 8]
```diff
@@ -87,7 +87,7 @@ typedef unsigned ErtsCodeIndex;
 typedef struct ErtsCodeMFA_ {
     Eterm module;
     Eterm function;
-    Uint  arity;
+    byte arity;
 } ErtsCodeMFA;
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply erts/emulator/beam/code_ix.h:91]: Yes, though it's a lot of code to go through and I'm not sure it's worth it.

**Comment by jhogberg** [comment]
> Looks promising. I wonder if there is a way to confirm the speed-up from less cache line trashing.

I don't expect this to speed anything up. At first it was done to scratch an itch, but after a small change to the `fcalls` field provoked a bug (https://github.com/erlang/otp/pull/7270) I started playing around with others to try to find more errors like it.

> Also, I wonder if reducing size of some variables could lead to unaligned access penalty with 64-bit CPUs.

I didn't change any packing attributes, so the performance should be unchanged as all fields are properly aligned according to their data size.

**Comment by michalmuskala** [comment]
Does this require an update to the efficiency guide: https://www.erlang.org/doc/efficiency_guide/advanced.html

> 
> Erlang process | 338 words when spawned, including a heap of 233 words.
> -- | --

**Comment by jhogberg** [comment]
Yes, and there's more to come there. :-)

**Pull request merged**